ENIGMAsystem/SHELL/Graphics_Systems/OpenGL1/GLmodel.cpp

```diff
@@ -309,54 +309,54 @@ bool d3d_model_load(int id, string fname)
 
 void d3d_model_part_draw(int id, int vertex_start, int vertex_count) // overload for no additional texture or transformation call's
 {
-    meshes[id]->Draw(vertex_start, vertex_count);
+  meshes[id]->Draw(vertex_start, vertex_count);
```

Comment:
again needs bounds check in debug mode

Reply:
This is the part I need to discuss with you or Josh. Due to the way I reimplemented the model functions, the way GMS does, this function can actually fail because I collapse triangle strips into indexed triangle lists sometimes (depending on which is arithmetically more efficient). What I mean is that, essentially, the vertex start may sometimes not be what the user expects because I have transformed the vertices they provided into something more efficient.

What I was actually considering was adding the even lower-level `vertex_*` functions and rewriting the model class with those (moving the model class to `Graphics_Systems/General` too). The vertex functions would not do any transformation to the vertices supplied (which I believe is how GMS works too) but the model classes will. That way models are fast and efficient, and vertex buffers very raw/low-level but provide you more power (but with more room to shoot yourself in the foot). Basically, the `vertex_` functions could let you draw in parts, but the model functions won't.

Then finally, I would rewrite the internal generified tiles to use the new `vertex_*` functions. This is going to be a complicated changeset, so I need to discuss it first. But right now, Direct3D9 still doesn't draw tiles because I can't sensible add `d3d_model_part_draw` to that system in a way that makes sense or reliable.

Reply:
That has nothing to do with the fact that you're accessing `meshes[id]` here without checking `id < meshes.size()` when debugging is active. I agree it needs to happen, and it's probably worth some design discussion, but checking your bounds in a debug macro just isn't gated on that.

Reply:
Sure, by extension of my above comments, I will add the debug macro when I generify the models.